src/Core/Services/MetadataProviders/SqlMetadataProvider.cs

```diff
@@ -1818,73 +1873,99 @@ IEnumerable<List<ForeignKeyDefinition>> foreignKeysForAllTargetEntities
-                // in the database, either from source->target or target->source entities or both.
-
-                // Add the referencing and referenced columns for this foreign key definition for the target.
+                // Check whether DAB found a match between 'configResolvedFkDefinition' and 'databaseResolvedFKDefinition's {referencing -> referenced} entity pair.
                 if (PairToFkDefinition is not null &&
-                    PairToFkDefinition.TryGetValue(fKDefinitionToTarget.Pair, out ForeignKeyDefinition? inferredFKDefinition))
+                    PairToFkDefinition.TryGetValue(configResolvedFkDefinition.Pair, out ForeignKeyDefinition? databaseResolvedFkDefinition))
                 {
-                    // Being here indicates that we inferred an FK constraint for the current foreign key definition.
-                    // The count of referencing and referenced columns being > 0 indicates that source.fields and target.fields 
-                    // have been specified in the config file.
-                    // In this scenario, higher precedence is given to the fields configured through the config file. So, the existing FK definition is retained as is.
-                    if (fKDefinitionToTarget.ReferencingColumns.Count > 0 && fKDefinitionToTarget.ReferencedColumns.Count > 0)
+                    if (DoesConfiguredRelationshipOverrideDatabaseFkConstraint(configResolvedFkDefinition))
                     {
-                        validatedFKDefinitionsToTarget.Add(fKDefinitionToTarget);
+                        validatedFKDefinitionsToTarget.Add(configResolvedFkDefinition);
+
+                        // Save additional metadata for use when processing requests on self-joined/referencing entities.
+                        if (IsSelfJoiningRelationship(configResolvedFkDefinition))
+                        {
+                            EntityRelationshipKey entityToFkDefKey = new(
+                                entityName: configResolvedFkDefinition.SourceEntityName,
+                                relationshipName: configResolvedFkDefinition.RelationshipName);
+                            RelationshipToFkDefinition.TryAdd(entityToFkDefKey, configResolvedFkDefinition);
+                        }
                     }
-                    // The count of referenced and referencing columns being = 0 indicates that source.fields and target.fields 
-                    // are not configured through the config file. In this case, the FK fields inferred from the database are populated.
                     else
                     {
-                        validatedFKDefinitionsToTarget.Add(inferredFKDefinition);
+                        validatedFKDefinitionsToTarget.Add(databaseResolvedFkDefinition);
+
+                        // Save additional metadata for use when processing requests on self-joined/referencing entities.
```

Comment:
Is this additional metadata not stored if its not a self joining relationship? Do we need to make it a special case?

Reply:
I intend that this is a special case because the additional metadata is only stored for self-joining relationships. My intention is that the new way of storing metadata is gradually adopted other types of relationships. This PR would be an order of magnitude larger  if I didn't special case the new metadata and saved it for every type of relationship + strip out whatever old relationship metadata we tracked.